### PR TITLE
Use bigints in F# build (#1053)

### PR DESCRIPTION
**Pull request by A-Manning**

### CODE REVIEW

**Comment by nikswamy** [comment]
Thanks for these improvements!

A couple of points: 

There seems to be more than one feature in the pull request. There's the fix to machine integer bounds checks, but there is also some stuff related to general extraction improvements for F#. Can you split this into two separate PRs, please? However, we are confused as to whether or not these extraction improvements have already been merged in by @cpitclaudel.

Now, as for the bounds check stuff: this looks pretty good overall. However, instead of duplicating the code between F# and OCaml in the already bloated FStar.Util.fs, how about we define a single interface: `FStar.Bigint` and provide two implementations of it. There are other places in the compiler that could make use of this abstraction.

In summary, let's have:

1. src/basic/boot/FStar.Bigint.fsi, which should be in the shared fragment of F# and F*
2. src/basic/boot/FStar.Bigint.fs, an implementation of the above interface in F# using .NET big ints.
3. src/basic/ml/FStar_Bigint.ml, which should be an implementation of this interface in OCaml using Zarith.

Thanks!

**Comment by A-Manning** [comment]
@nikswamy 
>There seems to be more than one feature in the pull request. There's the fix to machine integer bounds checks, but there is also some stuff related to general extraction improvements for F#.

Sorry about that, this seems to have been fixed by updating my branch to bring it up to date with master. This PR now only includes fixes to integer bounds.

Your proposal for `FStar.Bigint` is a better way to do things. I'll work on that and revise this PR.

**Comment by A-Manning** [comment]
@nikswamy 
>In summary, let's have:
>
> 1. src/basic/boot/FStar.Bigint.fsi, which should be in the shared fragment of F# and F*
> 2. src/basic/boot/FStar.Bigint.fs, an implementation of the above interface in F# using .NET big ints.
> 3. src/basic/ml/FStar_Bigint.ml, which should be an implementation of this interface in OCaml using Zarith.
>
>Thanks!

Done. A review of this PR at this point would be nice, but I'd like to work on using bigints in the type checker before a merge.

**Comment by catalin-hritcu** [comment]
This seems like a great fix. @wintersteiger Would you like to review this?

**Comment by wintersteiger** [comment]
Yes, except that I'm super busy until Wednesday, but after that I'd be happy to take a close look.

**Comment by nikswamy** [comment]
Looks reasonable at a quick glance. Thanks for the work @A-Manning. I will review in detail this week and hopefully merge.

File: src/basic/FStar.Const.fs [hunk 1]
```diff
@@ -47,27 +47,32 @@ let eq_const c1 c2 =
...
     let n =
         match width with
-        | Int8 -> 8
-        | Int16 -> 16
-        | Int32 -> 32
-        | Int64 -> 64
+        | Int8 -> big_int_of_string "8"
+        | Int16 -> big_int_of_string "16"
+        | Int32 -> big_int_of_string "32"
+        | Int64 -> big_int_of_string "32"
```
Comment (nikswamy) [inline src/basic/FStar.Const.fs]: typo

File: src/basic/basic.fsproj [hunk 2]
```diff
@@ -93,6 +93,8 @@
     <Compile Include="FStar.Options.fs" />
     <Compile Include="FStar.Unionfind.fsi" />
     <Compile Include="FStar.Unionfind.fs" />
+    <Compile Include="boot/FStar.BigInt.fsi" />
+    <Compile Include="boot/FStar.BigInt.fs" />
```
Comment (nikswamy) [inline src/basic/basic.fsproj]: FYI, all the files in the `boot` directory need to be adjacent to each other in the .fsproj file. Otherwise, visual studio will refuse to open it.

File: src/basic/ml/FStar_BigInt.ml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+let div_big_int = Z.div
+let mod_big_int = Z.rem
+
+let eq_big_int = Z.equal
+let le_big_int = Z.leq
+let lt_big_int = Z.lt
+let ge_big_int = Z.geq
+let gt_big_int = Z.gt
+
+let square_big_int = Z.sqrt
```
Comment (nikswamy) [inline src/basic/ml/FStar_BigInt.ml]: this looks like a typo too

File: src/basic/boot/FStar.BigInt.fs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+let div_big_int = Z.div_big_int
+let mod_big_int = Z.mod_big_int
+
+let eq_big_int = Z.eq_big_int
+let le_big_int = Z.le_big_int
+let lt_big_int = Z.lt_big_int
+let ge_big_int = Z.ge_big_int
+let gt_big_int = Z.gt_big_int
+
+let square_big_int = Z.sqrt_big_int
```
Comment (nikswamy) [inline src/basic/boot/FStar.BigInt.fs]: typo?

File: src/basic/ml/FStar_BigInt.ml (hunk 3 above)
Comment (nikswamy) [reply src/basic/ml/FStar_BigInt.ml]: Or at least a poorly named function

Comment (nikswamy) [reply src/basic/ml/FStar_BigInt.ml]: Changing the interface to expose `sqrt_big_int` since `square_big_int` can anyway be coded up by a client using `mult_big_int`.

File: src/reflection/FStar.Reflection.Data.fs [hunk 5]
```diff
@@ -39,7 +40,7 @@ type term_view =
     | Tv_Type   of unit
     | Tv_Refine of binder * term
     | Tv_Const  of vconst
-    | Tv_Uvar   of int * typ
+    | Tv_Uvar   of Z.t * typ
```
Comment (nikswamy) [inline src/reflection/FStar.Reflection.Data.fs]: This looks like overkill to me. It should just be an int, since we're only using ints in F# anyway to generate uvar ids.

Comment (nikswamy) [reply src/reflection/FStar.Reflection.Data.fs]: Although I suppose it makes the code more uniform

File: src/basic/ml/FStar_BigInt.ml (hunk 3 above)
Comment (A-Manning) [reply src/basic/ml/FStar_BigInt.ml]: Good catch, I've renamed it to `sqrt_big_int`

Comment (A-Manning) [reply src/basic/ml/FStar_BigInt.ml]: Good Catch! I've renamed it to `sqrt_big_int`.

**Comment by nikswamy** [comment]
Merged.

**Comment by nikswamy** [comment]
The F# version fails to parse hex numbers:

```
big_int_of_string failed to parse 0xffffffff
exception is System.FormatException: The value could not be parsed.
   at System.Numerics.BigNumber.ParseBigInteger(String value, NumberStyles style, NumberFormatInfo info)
   at FStar.BigInt.big_int_of_string@51.Invoke(String s)
Unexpected error
The value could not be parsed.
   at FStar.BigInt.big_int_of_string@51.Invoke(String s)
   at FStar.TypeChecker.Normalize.arg_as_bounded_int@688-1.Invoke(Tuple`2 tupledArg)
```

where

```
diff --git a/src/basic/boot/NotFStar.BigInt.fs b/src/basic/boot/NotFStar.BigInt.fs
index 9653c192f..80a367dfc 100644
--- a/src/basic/boot/NotFStar.BigInt.fs
+++ b/src/basic/boot/NotFStar.BigInt.fs
@@ -47,8 +47,13 @@ let gt_big_int = Z.gt_big_int
 let sqrt_big_int = Z.sqrt_big_int

 let string_of_big_int = Z.string_of_big_int
-let big_int_of_string = Z.big_int_of_string
-
+let big_int_of_string s =
+    try
+      Z.big_int_of_string s
+    with
+    e -> printfn "big_int_of_string failed to parse %s\nexception is %A"
+                s e;
+         raise e
 let of_int (x:int) = bigint x
 let to_int (x:bigint) = int32 x

```

**Comment by A-Manning** [comment]
@nikswamy  This seems to be caused by an incorrect implementation of `big_int_to_string` in `FSharp.Compatibility.Ocaml`. I have filed a bug report here https://github.com/fsprojects/FSharp.Compatibility/issues/18

**Comment by A-Manning** [comment]
@nikswamy Fixed in FSharp.Compatibility.Ocaml 0.1.14
